File: SortExamples/QuickSort.js
const generateArray = require('./Global').generateArray;

function quickSort(arr) {
    let left = 0,
        right = arr.length - 1;
    console.time('QuickSort');
    main(arr, left, right);
    console.timeEnd('QuickSort');
    return arr;
    function main(arr, left, right) {
        if(arr.length === 1) {
            return;
        }
        let index = partition(arr, left, right);
        if(left < index - 1) {
            main(arr, left, index - 1);
        }
        if(index < right) {
            main(arr, index, right);
        }
    }
    function partition(arr, left, right) {
        let pivot = arr[Math.floor((left + right) / 2)];
        while(left <= right) {
            while(arr[left] < pivot) {
                left++;
            }
            while(arr[right] > pivot) {
                right--;
            }
            if(left <= right) {
                [arr[left], arr[right]] = [arr[right], arr[left]];
                left++;
                right--;
            }
        }
        return left;
    }
}

console.log(quickSort([2, 1, 4, 2, 3, 6, 9, 8]));
// console.log(quickSort(generateArray(10000)));

File: SortExamples/PythonImplementation/QuickSort.py
from Global import generateList

def quickSort(l):
    def partition(l, left, right):
        mid = int((left + right) / 2)
        pivot = l[mid]
        while left <= right:
            while l[left] < pivot:
                left = left + 1
            while l[right] > pivot:
                right = right - 1
            if left <= right:
                tmp = l[left]
                l[left] = l[right]
                l[right] = tmp
                right = right - 1
                left = left + 1
        return left
    def main(l, left, right):
        if len(l) == 1:
            return l
        index = partition(l, left, right)
        if index - 1 > left:
            main(l, left, index - 1)
        if index < right:
            # 注意这里main函数的第一个参数不需要是截取后的list，那样的话下标会溢出。
            main(l, index, right)
        return l
    return main(l, 0, len(l) - 1)

print(quickSort(generateList(100)))

